### PR TITLE
Create template tool task

### PR DESCRIPTION
**Pull request by vpatakottu**
Create a Tool Task that invokes the SBOM CLI tool. This PR will only have an outline of the tool task without any implemented methods. It also has logic for conditionally compiling the GenerateSbomTask vs SbomCLIToolTask based on the target framework.

### CODE REVIEW

File: src/Microsoft.Sbom.Targets/GenerateSbom.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+using Microsoft.Build.Framework;
+
+/// <summary>
+/// This partial class defines and sanitizes the arguments that will be passed
+/// into the SBOM API and CLI tool for generation.
+/// </summary>
+public partial class GenerateSbom
+{
+    /// <summary>
+    /// The path to the drop directory for which the SBOM will be generated.
```
Comment (nschwerzler) [inline src/Microsoft.Sbom.Targets/GenerateSbom.cs:18]: Would add Gets or sets to the summary doc

Gets or sets the path to the drop directory...

File: src/Microsoft.Sbom.Targets/GenerateSbom.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    /// Version of the package the SBOM represents.
+    /// </summary>
+    [Required]
+    public string PackageVersion { get; set; }
+
+    /// <summary>
+    /// The base path of the SBOM namespace uri.
+    /// </summary>
+    [Required]
+    public string NamespaceBaseUri { get; set; }
```
Comment (nschwerzler) [inline src/Microsoft.Sbom.Targets/GenerateSbom.cs:45]: This should be a Uri to protect the input

File: src/Microsoft.Sbom.Targets/GenerateSbom.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    public string Verbosity { get; set; }
+
+    /// <summary>
+    /// A list of the name and version of the manifest format being used.
+    /// </summary>
+    public string ManifestInfo { get; set; }
+
+    /// <summary>
+    /// If true, it will delete the previously generated SBOM manifest directory before
+    /// generating a new SBOM in ManifestDirPath.
```
Comment (nschwerzler) [inline src/Microsoft.Sbom.Targets/GenerateSbom.cs:86]: Suggestion:
```
/// <summary>
/// Indicates whether the previously generated SBOM manifest directory should be deleted 
/// before generating a new SBOM in the directory specified by ManifestDirPath. 
/// Defaults to true.
/// </summary>
```

File: src/Microsoft.Sbom.Targets/Microsoft.Sbom.Targets.csproj [hunk 4]
```diff
@@ -39,6 +41,15 @@
     </ItemGroup>
   </Target>
 
+   <Target Name="AddSbomToolToPackage" AfterTargets="Build">
```
Comment (nschwerzler) [inline src/Microsoft.Sbom.Targets/Microsoft.Sbom.Targets.csproj:44]: We should figure out if there is a better AfterTargets for SBOM generation to only do this when they need an SBOM. Can do in a follow-up change once that is figured out.

File: src/Microsoft.Sbom.Targets/SbomCLIToolTask.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+/// </summary>
+public partial class GenerateSbom : ToolTask
+{
+    /// <summary>
+    /// Path to the SBOM CLI tool
+    /// </summary>
+    [Required]
+    public string SbomToolPath { get; set; }
+
+    protected override string ToolName => "Microsoft.Sbom.Tool";
```
Comment (nschwerzler) [inline src/Microsoft.Sbom.Targets/SbomCLIToolTask.cs:25]: Comment

File: src/Microsoft.Sbom.Targets/SbomCLIToolTask.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    /// <summary>
+    /// Get full path to SBOM CLI tool.
+    /// </summary>
+    /// <returns></returns>
+    protected override string GenerateFullPathToTool()
+    {
+        return $"{this.SbomToolPath}\\{this.ToolName}.exe";
```
Comment (nschwerzler) [inline src/Microsoft.Sbom.Targets/SbomCLIToolTask.cs:33]: Use Path.Combine to guard against extra slashes

File: src/Microsoft.Sbom.Targets/SbomCLIToolTask.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Sbom.Targets;
+
+using System;
```
Comment (nschwerzler) [inline src/Microsoft.Sbom.Targets/SbomCLIToolTask.cs:6]: Double check you need all these usings. Some look like they are extra.

File: src/Microsoft.Sbom.Targets/Microsoft.Sbom.Targets.csproj (hunk 4 above)
Comment (gustavoaca1997) [reply src/Microsoft.Sbom.Targets/Microsoft.Sbom.Targets.csproj:46]: This target is for including the SBOM CLI Tool into the Task's Nuget Package. I tried making the target run only during the packing step without success :(

File: src/Microsoft.Sbom.Targets/GenerateSbom.cs (hunk 1 above)
Comment (vpatakottu) [reply src/Microsoft.Sbom.Targets/GenerateSbom.cs]: Updated!

File: src/Microsoft.Sbom.Targets/GenerateSbom.cs (hunk 2 above)
Comment (vpatakottu) [reply src/Microsoft.Sbom.Targets/GenerateSbom.cs:45]: The MSBuild Task seems to only allow strings and bool's as parameters when we are passing them from the XML <Target> command, and the SBOM API accepts this value as a string as well. I believe the SBOM API does some URI checking, but should we implement some extra validation for this input as well?

File: src/Microsoft.Sbom.Targets/GenerateSbom.cs (hunk 3 above)
Comment (vpatakottu) [reply src/Microsoft.Sbom.Targets/GenerateSbom.cs]: Updated!

File: src/Microsoft.Sbom.Targets/SbomCLIToolTask.cs (hunk 5 above)
Comment (vpatakottu) [reply src/Microsoft.Sbom.Targets/SbomCLIToolTask.cs:15]: Moved this property to GenerateSbom.cs and updated the comment!

File: src/Microsoft.Sbom.Targets/SbomCLIToolTask.cs (hunk 6 above)
Comment (vpatakottu) [reply src/Microsoft.Sbom.Targets/SbomCLIToolTask.cs]: Updated!

**Review by gustavoaca1997 — APPROVED** [review]
Good job!

**Comment by vpatakottu** [comment]
@microsoft-github-policy-service agree company="Microsoft"

**Review by nschwerzler — APPROVED** [review]

**Pull request merged**